### ISSUE TITLE
documentation needed for audit logging

### ISSUE THREAD

**Issue by babbageclunk** [juju/docs#2435]
Audit logging has been rewritten - we used to have some code to do it but it wasn't really suitable, and it wasn't documented as far as I can tell. 

(I'm not sure how much detail to put into this, I haven't raised a docs bug before.)

The audit log records CLI commands by users that change things in the controller or models, and the API calls that are generated by those commands (as well as any errors that come back from the API).

It's driven by these controller config options (from [controller/config.go](https://github.com/juju/juju/blob/develop/controller/config.go#L34)):
- `auditing-enabled`
- `audit-log-capture-args`
- `audit-log-exclude-methods`
- `audit-log-max-backups`
- `audit-log-max-size`

It writes json log entries to `/var/log/juju/audit.log`, with file rotation controlled by the `max-size` and `-backups` options.

The log contains [records](https://github.com/juju/juju/blob/develop/core/auditlog/auditlog.go#L95) which are either:
- a [Conversation](https://github.com/juju/juju/blob/develop/core/auditlog/auditlog.go#L24), which corresponds to a top-level CLI command, 
- a [Request](https://github.com/juju/juju/blob/develop/core/auditlog/auditlog.go#L47) (a call to an API method)
- a [ResponseErrors](https://github.com/juju/juju/blob/develop/core/auditlog/auditlog.go#L70), which has details of any errors from calling the method.

I'm babbageclunk on irc as well if you need more details.

**Comment by pmatulis** [comment]
@babbageclunk Good to hear. I recall audit logging not working (an empty log file was left behind). I'll look into this. Thanks. :candy:

**Comment by pmatulis** [comment]
See PR #2128

**Comment by pmatulis** [comment]
@babbageclunk So is this for upcoming 2.4?

**Comment by pmatulis** [comment]
@babbageclunk I'm trying to come up with possible values for these options. Please confirm and fill in the missing piece.

- `auditing-enabled` : True | False
- `audit-log-capture-args` : True | False
- `audit-log-exclude-methods` : **?** (and some "methods" are excluded by default?)
- `audit-log-max-backups` : an integer (e.g. 2, 3, 4)
- `audit-log-max-size` : a size (e.g. 250M, 1G)

**Comment by babbageclunk** [comment]
@pmatulis Thanks! No, the full functionality will be in 2.3.3.

Oh, I meant to fill in values for those options! The ones you have there are all correct. 
`audit-log-exclude-methods` is a list of `Facade.Method` names that should be excluded. By default it has the special value `ReadOnlyMethods`, which corresponds to the following list of method names (none of which make changes):
```lang=go
var readonlyMethods = set.NewStrings(
	// Collected by running read-only commands.
	"Action.Actions",
	"Action.ApplicationsCharmsActions",
	"Action.FindActionsByNames",
	"Action.FindActionTagsByPrefix",
	"Application.GetConstraints",
	"ApplicationOffers.ApplicationOffers",
	"Backups.Info",
	"Client.FullStatus",
	"Client.GetModelConstraints",
	"Client.StatusHistory",
	"Controller.AllModels",
	"Controller.ControllerConfig",
	"Controller.GetControllerAccess",
	"Controller.ModelConfig",
	"Controller.ModelStatus",
	"MetricsDebug.GetMetrics",
	"ModelConfig.ModelGet",
	"ModelManager.ModelInfo",
	"ModelManager.ModelDefaults",
	"Pinger.Ping",
	"UserManager.UserInfo",

	// Don't filter out Application.Get - since it includes secrets
	// it's worthwhile to track when it's run, and it's not likely to
	// swamp the log.

	// All client facade methods that start with List.
	"Action.ListAll",
	"Action.ListPending",
	"Action.ListRunning",
	"Action.ListComplete",
	"ApplicationOffers.ListApplicationOffers",
	"Backups.List",
	"Block.List",
	"Charms.List",
	"Controller.ListBlockedModels",
	"FirewallRules.ListFirewallRules",
	"ImageManager.ListImages",
	"ImageMetadata.List",
	"KeyManager.ListKeys",
	"ModelManager.ListModels",
	"ModelManager.ListModelSummaries",
	"Payloads.List",
	"PayloadsHookContext.List",
	"Resources.ListResources",
	"ResourcesHookContext.ListResources",
	"Spaces.ListSpaces",
	"Storage.ListStorageDetails",
	"Storage.ListPools",
	"Storage.ListVolumes",
	"Storage.ListFilesystems",
	"Subnets.ListSubnets",
)
```
This list comes from [apiserver/observer/auditfilter.go](https://github.com/juju/juju/blob/2.3/apiserver/observer/auditfilter.go#L130) - it should be updated as we add new read-only methods. Users can exclude additional methods by setting the value to something like `[ReadOnlyMethods,Facade.OtherMethod]`.

Also, the option is called `audit-log-exclude-methods` but its behaviour is a bit more subtle than that. It doesn't just filter out API requests to those methods - that would mean that the log would still be full of  empty conversations for running `juju status`, making it harder to see interesting activity. Instead, any conversation that *only* contains calls to excluded methods is filtered out of the log. If a conversation contains calls to excluded methods and non-excluded methods, it will be in the log in full (even the call to excluded methods).
(I'm not sure whether you want to include this level of detail in the docs - maybe it's just confusing? - but if you do here it is.)

**Comment by pmatulis** [comment]
In your ```[ReadOnlyMethods,Facade.OtherMethod]```, what are the possible values for 'Facade.OtherMethod'?

**Comment by babbageclunk** [comment]
It could be the name of any API method that can be called in the Juju API. I'm not sure this is documented, sorry. I'd expect that the main place that a user would get values from is the audit log itself - if there was an API call in the log that they decided wasn't interesting, they'd it to the exclude list to remove it so that important messages are clearer.

**Comment by pmatulis** [comment]
@babbageclunk Alright, that's a pity. Now what about what happens in an HA context? Does the user need to check each controller? Like `logsink.log`?

**Comment by babbageclunk** [comment]
In an HA context, yes, you'd need to check the logs on all controller machines - each will have operations for that API server.

With respect to the Ping operations - yes, those are read-only, but they're part of a conversation where non-read-only methods are happening so they're included. You can see all of the messages in the conversation by grepping for the conversation-id `4e9dc976b17bfe88`. Those ones are a part of sshing into a machine, presumably to tail the audit log. :)

My thinking was: it's clearer to include all the messages in a conversation where something interesting was going on than to hide them - it makes it easier to understand what's happening. I discussed it with other people and they agreed, but maybe we were wrong. We're all devs who are thinking about the internals.

**Comment by pmatulis** [comment]
OK. And finally,

> If a conversation contains calls to excluded methods and non-excluded methods, it will be in the log in full (even the calls to excluded methods).

I'm guessing that this does not apply when 'ReadOnlyMethods' is part of the excluded filter? Any of those read-only methods will *never* be included?

**Comment by babbageclunk** [comment]
No, this applies even for the methods included in ReadOnlyMethods - if ModelManager.ListModels is called while executing `juju destroy-model mymodel` it will be included in the audit log (because of the ModelManager.Destroy call later on), but if the command run was `juju list-models` it won't appear.
